libraries/lwIP_PPP/src/PPPServer.cpp

```diff
@@ -0,0 +1,191 @@
+
+// This is still beta / a work in progress
+
+// testing on linux:
+// sudo /usr/sbin/pppd /dev/ttyUSB1 38400 noipdefault nocrtscts local defaultroute noauth nodetach debug dump
+// sudo /usr/sbin/pppd /dev/ttyUSB1 38400 noipdefault nocrtscts local defaultroute noauth
+
+// proxy arp is needed but we don't have it
+// http://lwip.100.n7.nabble.com/PPP-proxy-arp-support-tp33286p33345.html
+// using NAT instead (see in example)
+
+#include <Arduino.h>
+#include <Schedule.h>
+#include <IPAddress.h>
+#include <lwip/dns.h>
+
+#include "PPPServer.h"
+
+PPPServer::PPPServer(Stream* sio): _sio(sio), _cb(netif_status_cb_s), _enabled(false)
+{
+}
+
+bool PPPServer::handlePackets()
+{
+    size_t avail;
+    if ((avail = _sio->available()) > 0)
+    {
+        // XXX block peeking would be useful here
+        if (avail > _bufsize)
+        {
+            avail = _bufsize;
+        }
+        avail = _sio->readBytes(_buf, avail);
+        pppos_input(_ppp, _buf, avail);
+    }
+    return _enabled;
+}
+
+void PPPServer::link_status_cb_s(ppp_pcb* pcb, int err_code, void* ctx)
+{
+    (void)ctx;
+
+    netif* nif = ppp_netif(pcb);
+
+    switch (err_code)
+    {
+    case PPPERR_NONE:               /* No error. */
+    {
+#if LWIP_DNS
+        const ip_addr_t *ns;
+#endif /* LWIP_DNS */
+        ets_printf("ppp_link_status_cb: PPPERR_NONE\n\r");
+#if LWIP_IPV4
+        ets_printf("   our_ip4addr = %s\n\r", ip4addr_ntoa(netif_ip4_addr(nif)));
+        ets_printf("   his_ipaddr  = %s\n\r", ip4addr_ntoa(netif_ip4_gw(nif)));
+        ets_printf("   netmask     = %s\n\r", ip4addr_ntoa(netif_ip4_netmask(nif)));
+#endif /* LWIP_IPV4 */
+#if LWIP_IPV6
+        ets_printf("   our_ip6addr = %s\n\r", ip6addr_ntoa(netif_ip6_addr(nif, 0)));
+#endif /* LWIP_IPV6 */
+
+#if LWIP_DNS
+        ns = dns_getserver(0);
+        ets_printf("   dns1        = %s\n\r", ipaddr_ntoa(ns));
+        ns = dns_getserver(1);
+        ets_printf("   dns2        = %s\n\r", ipaddr_ntoa(ns));
+#endif /* LWIP_DNS */
+#if PPP_IPV6_SUPPORT
+        ets_printf("   our6_ipaddr = %s\n\r", ip6addr_ntoa(netif_ip6_addr(nif, 0)));
+#endif /* PPP_IPV6_SUPPORT */
+    }
+    break;
+
+    case PPPERR_PARAM:             /* Invalid parameter. */
+        ets_printf("ppp_link_status_cb: PPPERR_PARAM\n");
+        break;
+
+    case PPPERR_OPEN:              /* Unable to open PPP session. */
+        ets_printf("ppp_link_status_cb: PPPERR_OPEN\n");
+        break;
+
+    case PPPERR_DEVICE:            /* Invalid I/O device for PPP. */
+        ets_printf("ppp_link_status_cb: PPPERR_DEVICE\n");
+        break;
+
+    case PPPERR_ALLOC:             /* Unable to allocate resources. */
+        ets_printf("ppp_link_status_cb: PPPERR_ALLOC\n");
+        break;
+
+    case PPPERR_USER:              /* User interrupt. */
+        ets_printf("ppp_link_status_cb: PPPERR_USER\n");
```

Comment:
It can be useful here to remove the netif in order to give the user the possibility to start connection again without errors, like this:
```
netif_remove(&static_cast<PPPServer*>(ctx)->_netif);
```

Same could be done for connect errors.

Reply:
I forgot that `netif` is pulled out from `ppp_netif(pcb)` so it is possible another way:
```
netif_remove(nif);
```